### ISSUE TITLE
Proposal: Off-thread ViewModels

### ISSUE THREAD

**Issue by matthewp** [canjs/canjs#3739]
__tldr;__ this proposal is for a new library that allows you to have your ViewModels in a web worker while templates and components remain in the main thread.

Running non-UI code off-thread has the benefit of not blocking user interaction. Previously I worked on the [worker-render](https://github.com/canjs/worker-render) project which ran CanJS applications in a web worker. This strategy had a number of problems I explain below. This new proposal is to allow *ViewModels* to run in a web worker, while all view related code, like templates and event listeners, remain in the main thread.

# Problems

The [worker-render](https://github.com/canjs/worker-render) project was ultimately unsuccess for a couple of reasons. I would summarize them as:

* Workers do not have access to the DOM. Therefore to run code that expects a DOM you have to provide a fake one. We used can-simple-dom. While this covered most common cases in jQuery apps, it/is not complete. 
* Not all events can easily be handled off-thread, stuff like the `scroll` event.
* Debugging is more difficult, because devtools is able to display DOM nodes nicely, but not able to do so for vdom nodes.

Nevertheless, running *some* code in a worker has the benefit of not blocking painting, scrolling, and user events.

# Concept

The idea of this proposal is to allow ViewModels to exist in web workers, but keep things like templates and event handlers in the main thread. This means Models would be in a worker too.

From an API perspective, a new package would be created, let's call it `can-worker-observable`, that provides this capability. Here's an example of how this might look. In the main thread you would create a __proxy__ for your view model and pass it anywhere that viewmodels can be passed today:

__tabs.js__

```js
var Component = require("can-component");
var proxyViewModel = require("can-worker-observable").proxy;

var worker = new Worker("./worker.js");

var TabsViewModel = proxyViewModel(worker, "TabsViewModel");

Component.extend({
  tag: "tabs",
  ViewModel: TabsViewModel
});
```

And in your worker, you define the ViewModel:

__worker.js__

```js
var DefineMap = require("can-define/map/map");
var createProxy = require("can-worker-observable").create;

var Tabs = DefineMap.extend("TabsViewModel", { ... });

// createProxy uses the `fn.name` here to know what to sync with
createProxy(Tabs);
```

The library would handle creating new instances of the viewmodel within the worker and syncing it with the proxy that exists within the main thread.

**Comment by BigAB** [comment]
As a small suggestion on API, it seems like from the above example the `tabs.js` file could just as easily and more succinctly be

```
var Component = require("can-component");
var ProxyViewModel = require("can-worker-observable").Proxy;

var TabsViewModel = new ProxyViewModel("./tabs-worked.js");

Component.extend({
  tag: "tabs",
  ViewModel: TabsViewModel
});
```

**Comment by matthewp** [comment]
I haven't thought *too* much about the API yet, definite can use some work. The thing about your suggestion, though, is that it looks like it's one ViewModel per worker...

**Comment by BigAB** [comment]
Oh, would it not be?

I just assumed a worker would map to one viewModel class. but I supposed your hoping to import several ViewModels into the worked and then expose them on the `api` object.

**Comment by matthewp** [comment]
Yeah, that's the way this works as of now. Maybe this is too magical, could be something like `api.proxy("TabsViewModel")` or event `api.proxyConstructor("TabsViewModel")` to more easily support different types.

**Comment by rjgotten** [comment]
> Oh, would it not be?

Having a single worker per class is a bad idea because each individual worker involves memory overhead and contends for CPU resources in the scheduler. Ideally you never want to create more workers than there are physical cores.

There is also overhead involved in the actual creation of the workers and if you at one point rely on `importScripts` to load modules; that operation is _synchronous_ and blocking. Which means you need to set-up creation of workers from inside another worker, so as to not block the main thread.
